File: clojure/day19.clj
(ns day19
  (:require aoc
            [clojure.data.int-map :refer [dense-int-set]]))


(defrecord Blueprint [prices max-resources])


(defn line->blueprint [line]
  (let [[_ ore-ore clay-ore obs-ore obs-clay geode-ore geode-obs] line
        prices {:ore   {:ore ore-ore}
                :clay  {:ore clay-ore}
                :obs   {:ore obs-ore   , :clay obs-clay}
                :geode {:ore geode-ore , :obs geode-obs}}
        max-resources {:ore (max ore-ore clay-ore obs-ore geode-ore)
                       :clay obs-clay , :obs geode-obs , :geode 9999}]
    (->Blueprint prices max-resources)))


(defn best-case [resource bot time]
  (+ resource
     (* bot time)
     (/ (* time (inc time)) 2)))

(defn build-bot  [resources price]
  (reduce-kv
   (fn [rscs rsc cost]
     (assoc rscs rsc (- (resources rsc) cost)))
   resources
   price))

(defn no-negatives? [resources]
  (reduce-kv
   (fn [positive _ v]
     (if (neg? (long v))
       (reduced false)
       positive))
   true
   resources))


(defn calc-hash ^long [t bots resources]
  ;; "Hashing" function to avoid having a nested vector
  ;; as a hash-map key.
  ;;
  ;; Very fragile! Don't touch the coefficients,
  ;; i.e. I need a more robust way of doing this.
  (+ ^long t
     (* 33 (+ (* 2 ^long (:ore bots))
              (* 5 ^long (:clay bots))
              (* 13 ^long (:obs bots))
              (* 37 ^long (:geode bots))))
     (* 1151 (+ (* 3 ^long (:ore resources))
                (* 5 ^long (:clay resources))
                (* 11 ^long (:obs resources))
                (* 31 ^long (:geode resources))))))

(defn geodes [t bp]
  (let [types [:ore :clay :obs :geode]
        times {:ore 0 , :clay 4 , :obs 2 , :geode 0}
        state {:t t
               :bots      {:ore 1 , :clay 0 , :obs 0 , :geode 0}
               :resources {:ore 0 , :clay 0 , :obs 0 , :geode 0}
               :skipped   #{}}
        {:keys [prices max-resources]} bp
        ,
        collect (fn [resources bots t]
                  (->> resources
                       (reduce-kv (fn [acc res v]
                                    (assoc! acc res
                                            (min (* (max-resources res) t)
                                                 (+ v (bots res)))))
                                  (transient resources))
                       persistent!))]
    (loop [stack (list state)
           seen (transient (dense-int-set))
           score 0]
      (if-let [state (first stack)]
        (let [{:keys [t bots resources skipped]} state
              state-hash (calc-hash t bots resources)
              stack' (rest stack)]
          (if (or (zero? t)
                  (seen state-hash)
                  (< (best-case (resources :geode) (bots :geode) t) score))
            (recur stack' (conj! seen state-hash) (max score ^long (resources :geode)))
            (let [[stack'' skipped]
                  (reduce (fn [[stack can-build] bot]
                            (let [rscs (build-bot resources (prices bot))]
                              (if (and (not (skipped bot))
                                       (> t (times bot))
                                       (no-negatives? rscs)
                                       (< (bots bot) (max-resources bot))
                                       (< (resources bot) (* 1.5 (max-resources bot))))
                                [(conj stack {:t (dec t) , :resources (collect rscs bots t)
                                              :bots (update bots bot inc) , :skipped #{}})
                                 (conj can-build bot)]
                                [stack can-build])))
                          [stack' #{}]
                          types)
                  ,
                  stack
                  (if (or (< (resources :ore) (max-resources :ore))
                          (and (pos? (bots :clay))
                               (< (resources :clay) (max-resources :clay)))
                          (and (pos? (bots :obs))
                               (< (resources :obs) (max-resources :obs))))
                    (conj stack'' {:t (dec t) , :resources (collect resources bots t)
                                   :bots bots , :skipped skipped})
                    stack'')]
              (recur stack (conj! seen state-hash) score))))
        score))))

(defn part-1 [blueprints]
  (->> blueprints
       (pmap #(geodes 24 %))
       (mapv * (iterate inc 1))
       (reduce +)))

(defn part-2 [blueprints]
  (->> blueprints
       (take 3)
       (pmap #(geodes 32 %))
       (reduce *)))

(defn solve
  ([] (solve (aoc/read-file 19)))
  ([input]
   (let [blueprints (aoc/parse-input input (comp line->blueprint aoc/integers))
         p1 (future (part-1 blueprints))
         p2 (future (part-2 blueprints))]
     [@p1 @p2])))


(time (solve))

File: python/day19.py
from aoc import *
from collections import namedtuple
from math import prod


Bots = namedtuple('Bots', ('ore', 'clay', 'obs', 'geode'))
Resources = namedtuple('Resources', ('ore', 'clay', 'obs', 'geode'))


def best_case(resource, bot, time):
    return resource + time*bot + time*(time+1)//2


def geodes(costs, t=24,
           bots=Bots(1, 0, 0, 0),
           resources=Resources(0, 0, 0, 0)):
    score = 0
    seen = set()
    ore_ore, clay_ore, obs_ore, obs_clay, geode_ore, geode_obs = costs
    max_ore = max(ore_ore, clay_ore, obs_ore, geode_ore)
    max_clay = obs_clay
    max_obs = geode_obs
    stack = [(t, bots, resources, set())]
    while stack:
        t, bots, resources, skipped = state = stack.pop()
        if (state[:-1] in seen
                or t == 0
                or best_case(resources.geode, bots.geode, t) < score):
            score = max(score, resources.geode)
            continue
        seen.add(state[:-1])
        t -= 1
        ore = min(max_ore*t, resources.ore + bots.ore)
        clay = min(max_clay*t, resources.clay + bots.clay)
        obs = min(max_obs*t, resources.obs + bots.obs)
        geode = resources.geode + bots.geode
        can_build = set()
        if (t > 0 and 'geode' not in skipped
            and resources.ore >= geode_ore and resources.obs >= geode_obs):
            can_build.add('geode')
            stack.append((t,
                          bots._replace(geode=bots.geode+1),
                          Resources(ore-geode_ore, clay, obs-geode_obs, geode),
                          set()))
            # continue   # faster with this, but people say it would be "cheating"
        if (t > 2 and 'obs' not in skipped
            and resources.ore >= obs_ore and resources.clay >= obs_clay
            and bots.obs < max_obs and resources.obs < 1.2*max_obs):
            can_build.add('obs')
            stack.append((t,
                          bots._replace(obs=bots.obs+1),
                          Resources(ore-obs_ore, clay-obs_clay, obs, geode),
                          set()))
        if (t > 4 and 'clay' not in skipped
            and resources.ore >= clay_ore
            and bots.clay < max_clay and resources.clay < 1.2*max_clay):
            can_build.add('clay')
            stack.append((t,
                          bots._replace(clay=bots.clay+1),
                          Resources(ore-clay_ore, clay, obs, geode),
                          set()))
        if ('ore' not in skipped and resources.ore >= ore_ore
            and bots.ore < max_ore and resources.ore < 1.5*max_ore):
            can_build.add('ore')
            stack.append((t,
                          bots._replace(ore=bots.ore+1),
                          Resources(ore-ore_ore, clay, obs, geode),
                          set()))
        if (resources.ore < max_ore or (bots.clay and resources.clay < max_clay)
            or (bots.obs and resources.obs < max_obs)):
            stack.append((t,
                          bots,
                          Resources(ore, clay, obs, geode),
                          can_build))
    return score


def part_1(blueprints):
    return sum(bp * geodes(costs)
               for bp, *costs in blueprints)


def part_2(blueprints):
    return prod(geodes(costs, 32)
                for _, *costs in blueprints[:3])


def solve(filename=19):
    blueprints = mapl(integers, read_input(filename))
    return part_1(blueprints), part_2(blueprints)


print(solve())










def test():
    assert solve("19_test") == (33, 3472)
    assert solve() == (994, 15960)
